src/Chapter4.hs

```diff
@@ -488,7 +496,18 @@ Implement the 'Applicative' instance for our 'List' type.
   may also need to implement a few useful helper functions for our List
   type.
 -}
+instance Applicative List where
+  pure :: a -> List a
+  pure x = Cons x Empty
 
+  (<*>) :: List (a -> b) -> List a -> List b
+  Empty <*> _ = Empty
+  (Cons f Empty) <*> xs = fmap f xs
+  (Cons f fs) <*> xs = join (fmap f xs) (fs <*> xs)
+
+join :: List a -> List a -> List a
+join Empty xs = xs
```

Comment:
Perfect!

You can one more case to make your solution even more efficient:

```haskell
join xs Empty = xs
```

In that case, you won't need to walk through each element of the first list when the second one is empty and return the list instantly 🚤 

Also `join` is the name of a monadic function (that inlines the monad in the same monad). Haskell allows name shadowing, that's why it works. Though, in practice, it's usually better to avoid name clashes with other functions cause it could be confusing 🙂
The standard operation of lists appending, so you can name it `append` or anything else you would prefer! 👍🏼